moabb/evaluations/splitters.py

```diff
@@ -0,0 +1,133 @@
+
+from moabb.evaluations.metasplitters import PseudoOnlineSplit
+
+
+class WithinSessionSplitter(BaseCrossValidator):
+    """Data splitter for within session evaluation.
+
+    Within-session evaluation uses k-fold cross_validation to determine train
+    and test sets for each subject in each session. This splitter
+    assumes that all data from all subjects is already known and loaded.
+
+    .. image:: images/withinsess.png
+        :alt: The schematic diagram of the WithinSession split
+        :align: center
+
+
+    Parameters
+    ----------
+    n_folds : int
+        Number of folds. Must be at least 2. If
+    random_state: int, RandomState instance or None, default=None
+        Controls the randomness of splits. Only used when `shuffle` is True.
+        Pass an int for reproducible output across multiple function calls.
+    shuffle : bool, default=True
+        Whether to shuffle each class's samples before splitting into batches.
+        Note that the samples within each split will not be shuffled.
+    custom_cv: bool, default=False
+        Indicates if you are using PseudoOnlineSplit as cv strategy
+    calib_size: int, default=None
+        Size of calibration set if custom_cv==True
+    cv: cros-validation object, default=StratifiedKFold
+        Inner cross-validation strategy for splitting the sessions. Be careful, if
+        PseudoOnlineSplit is used, it will return calibration and test indexes.
+
+
+    Examples
+    -----------
+
+    >>> import pandas as pd
+    >>> import numpy as np
+    >>> from moabb.evaluations.splitters import WithinSessionSplitter
+    >>> X = np.array([[1, 2], [3, 4], [5, 6], [1,4], [7, 4], [5, 8], [0,3], [2,4]])
+    >>> y = np.array([1, 2, 1, 2, 1, 2, 1, 2])
+    >>> subjects = np.array([1, 1, 1, 1, 1, 1, 1, 1])
+    >>> sessions = np.array(['T', 'T', 'T', 'T', 'E', 'E', 'E', 'E'])
+    >>> metadata = pd.DataFrame(data={'subject': subjects, 'session': sessions})
+    >>> csess = WithinSessionSplitter(n_folds=2)
+    >>> csess.get_n_splits(metadata)
+    4
+    >>> for i, (train_index, test_index) in enumerate(csess.split(y, metadata)):
+    ...    print(f"Fold {i}:")
+    ...    print(f"  Train: index={train_index}, group={subjects[train_index]}, session={sessions[train_index]}")
+    ...    print(f"  Test:  index={test_index}, group={subjects[test_index]}, sessions={sessions[test_index]}")
+    Fold 0:
+      Train: index=[4 7], group=[1 1], session=['E' 'E']
+      Test:  index=[5 6], group=[1 1], sessions=['E' 'E']
+    Fold 1:
+      Train: index=[5 6], group=[1 1], session=['E' 'E']
+      Test:  index=[4 7], group=[1 1], sessions=['E' 'E']
+    Fold 2:
+      Train: index=[2 3], group=[1 1], session=['T' 'T']
+      Test:  index=[0 1], group=[1 1], sessions=['T' 'T']
+    Fold 3:
+      Train: index=[0 1], group=[1 1], session=['T' 'T']
+      Test:  index=[2 3], group=[1 1], sessions=['T' 'T']
+    """
+
+    def __init__(
+        self,
+        cv=StratifiedKFold,
```

Comment:
In the scikit-learn framework, `cv` is a cross-validator object, not a class. I think it would be best to stick to it. This would avoid to instantiate it during the `split` call. You can have `split=None` by default and instantiate `cv=StratifiedKFold()` class in the `__init__`  method.
Also, you can check the cf argument with sklearn’s `check_cv`

Reply:
I understand the concern, but I'm a bit unsure on how to implement it in the case shuffle=True, since I'm defining a different seed for each (subject, session). The suggestion is to instantiate cv in the __init__ method in case split is not needed, and keep how it is being done otherwise?

Reply:
Indeed it's not easy. But you could for example make a wrapper around `StratifiedKFolds` which would instantiate a different cv with a different seed for each subject/session.

Also I just noticed that at the moment, the seeds for each subject/session are chosen at random. We will not be able to have reproducible results this way. Instead, you could add a parameter `global_seed` to your wrapper and use, for each cv, `random_state = global_seed + 10000*subject_number + session_number` (it's safe to say we will never have 10000 sessions) if `global_seed` is an integer and none otherwise